Update samrai 2.4.4

drwells opened the pull request:
I know it seems a bit unusual to add a patch supporting this ancient version of SAMRAI (released in 2008) but we need it for a project here at UNC (www.github.com/ibamr/ibamr). This version of SAMRAI has a bunch of (easy to fix) ADL problems and some performance issues; the provided patch addresses both.

adamjstewart commented on var/spack/repos/builtin/packages/samrai/package.py:
```diff
@@ -60,19 +60,26 @@ class Samrai(AutotoolsPackage):
-    patch('no-tool-build.patch', when='%gcc')
+    # Versions after 3.0 but and before (but not including) 3.12.0 depend on
+    # boost. Versions 3.0 - 3.11.5 can only work with boost version 1.64.0 or
+    # earlier.
```
This comment is pretty useless. It just translates the line below to English. Not sure if it is necessary.

drwells replied:
Fixed.

adamjstewart commented on var/spack/repos/builtin/packages/samrai/package.py:
```diff
@@ -89,6 +96,10 @@ def configure_args(self):
             '--with-petsc=no'])
 
+        # SAMRAI 2 used templates; enable implicit instantiation
+        if self.version == Version('2.4.4'):
```
I know 2.4.4 is the only 2 currently in Spack, but just in case someone adds another, can you change this to:
```python
if self.spec.satisfies('@:2'):
```

drwells replied:
You meant `'@:3'`, right?

adamjstewart commented on var/spack/repos/builtin/packages/samrai/package.py:
```diff
@@ -98,16 +109,20 @@ def configure_args(self):
+            options.append('--with-silo=%s' % self.spec['silo'].prefix)
+
+        if Version('3.0.0') <= self.version:
+            if Version('3.12.0') <= self.version:
```
Combine these into a single line:
```python
if self.spec.satisfies('@3.0:3.12'):
```

drwells replied:
I reorganized this whole block to use that syntax and also `setup_dependent_environment`.

drwells commented:
Thanks for reviewing this. I addressed all review comments and cleaned up the version checks.

AFAICT `setup_dependent_environment` is the best way to set up `CXXFLAGS`; I used that to keep things organized. Please let me know if there is a better way to append flags.

drwells commented:
Please let me know if there is anything else that could be improved in this PR. I would like to get this merged before it gets stale :)

scheibelp commented:
Thanks!
